lib/connection.js

```diff
@@ -707,6 +707,10 @@ class Connection extends EventEmitter {
     return this.addCommand(new Commands.Ping(cb));
   }
 
+  reset(cb) {
```

Comment:
I believe this need to perform some additional client cleanup, for example prepared statements need to be discarded.

Reply:
ok let me take a look

Reply:
Preparedstatement is discarded automatically
https://mariadb.com/kb/en/com_reset_connection/

Reply:
yes, server side, but we need to delete statements cache otherwise it might be used incorrectly with `execute()` after reset

I think this is what needs to be done

1) add private connection helper method `connection._initialiseStatementCache({ disableDispose: boolean })`

something along the lines:

```js
_initialiseStatementCache({ disableDispose }) {
  if (this._statements) {
    if (disableDispose) {
       this._statements.dispose = null;
    }
    this._statements.clear();
  }
  this._statements = new LRU({
    max: this.config.maxPreparedStatements,
    dispose: function(key, statement) {
      statement.close();
    }
  })
}
```

2. replace line [66 in connection.js](https://github.com/sidorares/node-mysql2/blob/299ed3d95576da72ce71933ea26cb76f8ab21f79/lib/connection.js#L66) with `this._initialiseStatementCache({ disableDispose: false });`

3) in `ResetConnectionCommand.resetConnectionResponse(packet, connection)` add call `connection._initialiseStatementCache({ disableDispose: true })` [at the beginning](https://github.com/sidorares/node-mysql2/blob/299ed3d95576da72ce71933ea26cb76f8ab21f79/lib/commands/reset_connection.js#L19) before calling `onResult()`

Reply:
Ah... got it

Reply:
I thought statement can be reused even though the connection is reset

Reply:
I need to double check but I'm pretty sure statement id is sequential 1, 2, 3, ... for each `prepare()` command ( can't remember if 0 or 1 based ) and after reset it goes again as 1, 2, 3, ...

Reply:
small correction: set dispose to noop instead of null, e.i `this._statements.dispose = () => {};`

Reply:
gotcha